### PR TITLE
fix: Won't build tests & docs if not top project.

### LINKED ISSUE #4: [Bug] Compilation failure on the repo tests preventing its usage
First of all, thanks for this awesome work! This library is really interesting!

## The Bug

When trying to build a minimal sample on Windows (latest Visual Studio compiler) using your library through `CMake` `FetchContent` I got stuck on building your tests instead of mine (compilation fails due some tuple magic not succeeding apparently).

## System Details

 - OS: Windows 11
 - Compiler version: MSVC 19.35.32217.1 (Visual Studio 17 2022 Windows SDK version 10.0.22621.0 to target Windows 10.0.25357)
 - Version / commit of Clean Test: d994ea4e3c28177bd763af9dacb9533d554bc7e8

## Reproduce

One CMakeLists.txt file

```cmake
cmake_minimum_required(VERSION 3.17)
project(clean-test-win)

set(CMAKE_CXX_STANDARD 20)

include(FetchContent)
fetchcontent_declare(
  CleanTest
  GIT_REPOSITORY https://github.com/clean-test/clean-test.git
  GIT_TAG        main # or a fixed tag e.g. v1.2.0
)
fetchcontent_makeavailable(CleanTest)

add_executable(clean-test-win test.cpp)
target_link_libraries(clean-test-win CleanTest::main-automatic)
```

and one C++ source file:

```cpp
#include <clean-test/clean-test.h>

constexpr auto sum(auto... vs) { return (0 + ... + vs); }

namespace ct = clean_test;
using namespace ct::literals;

auto const suite = ct::Suite{ "sum", [] {
    "0"_test = [] { ct::expect(sum() == 0_i); };
    "3"_test = [] { ct::expect(sum(1, 2) == 1 + 2_i); };
    "A"_test = [] { ct::expect(sum(-1) < 0_i and sum(+1) > 0_i); };
} };
```

Then the regular CMake workflow:

```sh
cmake -E make_directory _build

cmake -S . -B _build

```
## Observed Behavior

Build fails without even reaching my target (it stopped on failing to compile the library tests on some tuple related concept matching).

## Expected Behavior

By default build only the targets we need in order to use the library.

## Additional Context

The complete build log is available in: https://github.com/CarlosNihelton/CleanTestWin/blob/main/log.txt.
The repository where that file lives also contain the CMakeLists.txt and the C++ source file described above.

I could complete my build successfully and use your library as expected by skipping your tests with `cmake -S . -B _build _DCLEANTEST_TEST=OFF`.

Although it wouldn't solve the real issue, moving the test and documentation build steps into the section which is skipped if `clean-test` is not the top level project, also makes it, with the advantage of making it easy for newcomers to use your library by default. 

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 427fc37..dcbe2ba 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -234,16 +234,17 @@ else ()
     add_library(CleanTest::main-automatic ALIAS cleantest-main-shared)
 endif ()

-if (CLEANTEST_TEST)
-    enable_testing()
-    add_subdirectory(test)
-endif()
-add_subdirectory(doc)
-
-## Packaging ###########################################################################################################
+## Packaging, doc and tests ###########################################################################################################

 # Skip if we are not the toplevel of the cmake build
 if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
+
+    if (CLEANTEST_TEST)
+        enable_testing()
+        add_subdirectory(test)
+    endif()
+    add_subdirectory(doc)
+
     set(CPACK_PACKAGE_NAME                clean-test)
     set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "A modern, C++-20 unit-testing framework.")
     set(CPACK_PACKAGE_INSTALL_DIRECTORY   ${CPACK_PACKAGE_NAME})
```

Consider the case of maintainers of a vcpkg port or a conan recipe, for instance, which would benefit from that change by finding their intended usage by default, without the need to tweak CMake variables.

### PR DESCRIPTION
**Pull request by cnihelton**
## References

 - Issue: Fixes #4.

## What changes

Moves building docs and tests into the same section as packaging, to run only if this is the top level project.

## Checklist

- [x] Bug fix (change which fixes issues)
- [ ] New feature (change which introduces functionality)
- [ ] Breaking change (that causes any existing test to change, e.g. API / ABI change)
- [x] I have read the [contributing guidelines](CONTRIBUTING.md).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by m8mble** [comment]
Thanks for the PR. With your change, users can no longer enable tests when they use `clean-test` in a sub-directory - even if they want to. Professional CMake (Ed 15) suggests the following:

```cmake
option(CLEANTEST_TEST "Build tests for the framework." ${PROJECT_IS_TOP_LEVEL})

# [ ... ]

if (CLEANTEST_TEST)
  if (PROJECT_IS_TOP_LEVEL)
    enable_testing()
  endif()
  add_subdirectory(test)
endif()
if (PROJECT_IS_TOP_LEVEL)
  # [ ... ] (packaging
else()
  mark_as_advanced(CLEANTEST_TEST)
endif()
```
Would that meet your expectations?

Similar considerations apply to the `docs` folder. If docs are annoying by default, I'd prefer to rather exclude related targets from `all` but keeping the `add_subdirectory(docs)` unconditionally.

**Comment by CarlosNihelton** [comment]
> Would that meet your expectations?

Absolutely! `PROJECT_IS_TOP_LEVEL` requires CMake 3.21 or higher, though. 

> 
> Similar considerations apply to the `docs` folder. If docs are annoying by default, I'd prefer to rather exclude related targets from `all` but keeping the `add_subdirectory(docs)` unconditionally.

Indeed, the `add_custom_target` doesn't affect the `ALL` target by default, so since you didn't add the  `ALL` flag when adding the doc targets, we are fine with unconditionally adding the docs sub-directory.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Copyright (c) m8mble 2020.
 # SPDX-License-Identifier: BSL-1.0
 
-cmake_minimum_required(VERSION 3.17)
+cmake_minimum_required(VERSION 3.21)
```
Comment (m8mble) [inline CMakeLists.txt:4]: Thanks for spotting this. Let me suggest
```cmake
set(CLEANTEST_IS_TOP_LEVEL OFF)
if (CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
    set(CLEANTEST_IS_TOP_LEVEL ON)
endif()
```
(and replacing `PROJECT_IS_TOP_LEVEL => CLEANTEST_IS_TOP_LEVEL`) as a workaround in order to stay with the `cmake` version.

File: CMakeLists.txt [hunk 2]
```diff
@@ -285,4 +287,6 @@ if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
         DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
     )
     include(CPack)
+else()
+  mark_as_advanced(CLEANTEST_TEST)
```
Comment (m8mble) [inline CMakeLists.txt:291]: :+1:

**Review by m8mble — COMMENTED** [review]
Thanks for checking the minimum required `cmake` version.

**Review by m8mble — APPROVED** [review]
LGTM

**Comment by m8mble** [comment]
Thank you very much for fixing this.

**Comment by CarlosNihelton** [comment]
Thank you very much for this awesome library! I'd like to see it widespread.

**Pull request merged**